cleanlab/dataset.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (C) 2017-2022  Cleanlab Inc.
+# This file is part of cleanlab.
+#
+# cleanlab is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+
+"""Dataset Module
+Supports dataset-level and class-level automated quality, including finding which classes should
+be merged with other classes in your dataset, which classes should be removed,
+and which classes have the best label accuracy, and which have the worst labels.
+"""
+
+import numpy as np
+from count import estimate_joint, compute_confident_joint
+
+
+def find_similar_classes(
+    joint=None,
+    asymmetric=True,
+    *,
+    labels=None,
+    pred_probs=None,
+    confident_joint=None,
+    multi_label=False,
+):
+    """Ranks all pairs of classes by their similarity and returns the ordered pairs as well as
+    a similarity score between 0 and 1. Use this function to determine which classes in a dataset
+    should be merged.
+
+    The method uses the joint distribution of noisy and true labels to compute ontological issues
+    via the approach published in (Northcutt, Jiang, Chuang (2021) JAIR).
+    paper: https://arxiv.org/abs/1911.00068
+
+    This method works by providing any one (and only one) of the following inputs:
+    1. confident_joint
+    2. joint
+    3. labels and pred_probs
+    but it is not necessary to provide a combination, e.g. the labels and pred_probs and joint.
+
+    For parameter info, see the docstring of `count.estimate_joint`
+
+    Parameters
+    ----------
+    See `count.estimate_joint` docstring.
+
+    asymmetric : bool (default: True)
+      If `asymmetric==True`, includes both pairs (class1, class2) and (class2, class1). Use this for
+      finding "is a" relationships where for example "class1 is a class2".
+      If `asymmetric==False`, the pair (class1, class2) will onyl be returned once and order is
+      arbitrary (internally this is just summing score(class1, class2) + score(class2, class1)).
+
+    Returns
+    -------
+        A tuple of 2 ordered lists: (list_of_pairs_of_classes, likelihood_they_should_be_merged)"""
```

Comment:
likelihood_they_should_be_merged is not well-defined.
I think it may be overly bold to claim "should_be_merged".

Two classes could be fundamentally confusing to annotators while still not mergeable (eg. "benign tumor" vs "malignant tumor" in medical images).

Reply:
What is advantage of this return type vs say:
List[tuple] where each tuple = (class1, class2, confusion_score)

Or pd.DataFrame outlined in above comment.

Reply:
the confusion_score (what you're calling likelihood_they_should_be_merged) should have some intuitive definition of what it's value represents right?

Reply:
Agree. Any recommendations? It's based on the joint probability.

Reply:
Hm... as a user the information that I'd be most interested in is:
234 examples labeled as class A probably should actually be labeled as class B instead.

So I'd print that at least. 
If scores  = joint probability, you could explain it as:
"score(A,B) = The proportion of examples in the dataset that are labeled as class A but probably should actually be labeled as class B instead."

Reply:
I implemented conditional probability scores for the class level functionality and added thorough explanations in all Returns sections